### PR TITLE
o/configstate/configcore: support snap set system swap.size=...

### PR DESCRIPTION
**Pull request by anonymouse64**
Add a handler for the system swap option which manipulates /etc/default/swapfile
and swapfile.service. This option works both with systemd around at runtime, and
without during the initrd for example when only the rootfs is available.

### CODE REVIEW

File: overlord/configstate/configcore/swap.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		return err
+	}
+
+	if opts == nil {
+		// if we are not doing this filesystem only, then we need to also
+		// restart the swap service
+		sysd := systemd.NewUnderRoot(dirs.GlobalRootDir, systemd.SystemMode, &backlightSysdLogger{})
+
+		// TODO: what's an appropriate amount of time to wait here?
+		if err := sysd.Restart("swapfile.service", 60*time.Second); err != nil {
```
Comment (ogra1) [inline overlord/configstate/configcore/swap.go:131]: on my UC20 this is a disabled service by default, does *.Restart() also enable it ?

Comment (anonymouse64) [reply overlord/configstate/configcore/swap.go:131]: Well AFAIR, the service just calls swapon swapoff, etc. so it doesn't need to be running persistently. I did test it locally on UC20 and it works after a reboot, etc.

**Review by mvo5 — APPROVED** [review]
This looks good and useful to me. I don't think we let people configure swap on uc20 systems right now so this is a useful addition for these systems.

File: overlord/configstate/configcore/swap.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	// ensure the directory exists
+	if err := os.MkdirAll(filepath.Dir(swapConfigPath), 0755); err != nil {
+		return err
+	}
+
+	// the size of swap needs to be specified in Megabytes, while quantity.Size
+	// is a uint64 of bytes
+	fileContent := fmt.Sprintf("FILE=%s\nSIZE=%d\n", location, szBytes/quantity.SizeMiB)
```
Comment (bboozzoo) [inline overlord/configstate/configcore/swap.go:118]: We should probably have a check somewhere earlier that `szBytes >= quantity.SizeMiB && szBytes % quantity.SizeMiB == 0`

**Comment by pedronis** [comment]
swapfile.service is provided by core*, not something preexisting?   from a naming POV I see 3 paths:

* swap.size
* system.swap.size   even if it's something we provide,  to match the fact that we have system.journal...
* something that mentions  "vm"?  though on the  sysctl side  vm.swappiness is the only thing explicitly related to swap

File: overlord/configstate/configcore/swap.go (hunk 2 above)
Comment (anonymouse64) [reply overlord/configstate/configcore/swap.go:135]: I added these checks

**Comment by anonymouse64** [comment]
@pedronis yes swapfile.service comes from core* 

Re the naming, I don't see the difference between your two proposals, from a user standpoint I wrote it like this to match journal setting:


```
$ snap set system swap.size=512M
```

which matches what we do with journal:
```
$ snap set system journal.persistent=true
```

Or did I misunderstand how things work? FWIW, we don't have a top level key system for snap options:

```
$ sudo snap set system system.journal.persistent=true
error: cannot perform the following tasks:
- Run configure hook of "core" snap (run hook "configure": cannot set "core.system.journal.persistent": unsupported system option)
```

I don't think including VM is useful since as you mention the only related thing is vm.swappiness is the only VM related thing, and swap has a pretty standard usage and is understandable to the layperson.

**Comment by pedronis** [comment]
> @pedronis yes swapfile.service comes from core*
> 
> Re the naming, I don't see the difference between your two proposals, from a user standpoint I wrote it like this to match journal > setting:
>
> $ snap set system swap.size=512M
> which matches what we do with journal:
>
> $ snap set system journal.persistent=true

apparently I misread the code (not sure how/why)

I think the name is fine then. I don't think this need my close review apart the naming

**Comment by pedronis** [comment]
I removed myself as needed as the option name is fine as is.

**Comment by anonymouse64** [comment]
This PR is fully ready now with unit tests

File: overlord/configstate/configcore/swap.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/mvo5/goconfigparser"
+	"github.com/snapcore/snapd/dirs"
```
Comment (bboozzoo) [inline overlord/configstate/configcore/swap.go:30]:
```suggestion
	"github.com/mvo5/goconfigparser"

	"github.com/snapcore/snapd/dirs"
```

**Review by bboozzoo — APPROVED** [review]
LGTM. It'd be great to check the timings on a real device.

File: overlord/configstate/configcore/swap.go (hunk 1 above)
Comment (anonymouse64) [reply overlord/configstate/configcore/swap.go:148]: yeah indeed it does, in a VM it took about 30 seconds to execute `snap set system swap.size=4G`

File: overlord/configstate/configcore/handlers.go [hunk 4]
```diff
@@ -85,6 +85,9 @@ func init() {
 	// system.disable-backlight-service
 	addFSOnlyHandler(validateBacklightServiceSettings, handleBacklightServiceConfiguration, coreOnly)
 
+	// swap.size
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply overlord/configstate/configcore/handlers.go:88]: actually this is wrong, that's only used for options which are themselves prefixed with system, such as disable-backlight-service:
```
$ sudo snap get -d system
{
	"cloud": {
		"name": "unknown"
	},
	"journal": {
		"persistent": true
	},
	"refresh": {},
	"seed": {
		"loaded": true
	},
	"swap": {
		"size": "1G"
	},
	"system": {
		"disable-backlight-service": true
	}
}
```

**Comment by bboozzoo** [comment]
The failures are unrelated. @pedronis @mvo5 can you land it?

**Review by mvo5 — APPROVED** [review]
Thank you

**Pull request merged**